api/controllers/build.js

```diff
@@ -30,38 +30,41 @@ const emitBuildStatus = build => Site.findByPk(build.site)
     socketIO.to(siteRoom).emit('build status', msg);
     const builderRoom = SocketIOSubscriber.getBuilderRoom(build.site, build.user);
     socketIO.to(builderRoom).emit('build status', msg);
-    return Promise.resolve();
-  })
-  .catch(err => logger.error(err.stack));
+  } catch (err) {
+    EventCreator.error(Event.labels.SOCKET_IO, err, { buildId: build.id });
+  }
+};
 
 const saveBuildToProxy = async (build) => {
   const buildSettings = await GithubBuildHelper.fetchContent(build, BUILD_SETTINGS_FILE);
   if (buildSettings) {
     const settings = JSON.parse(buildSettings);
     await ProxyDataSync.saveBuild(build, settings);
-    EventCreator.audit(Event.labels.BUILD_STATUS, build, `${BUILD_SETTINGS_FILE} saved to proxy database`);
+    const body = {
+      filename: BUILD_SETTINGS_FILE,
+      commitSha: build.clonedCommitSha,
+    };
+    const message = `${BUILD_SETTINGS_FILE} saved to proxy database`;
+    EventCreator.audit(Event.labels.PROXY_EDGE, build, message, body);
   }
 };
 
-module.exports = {
-  find: (req, res) => {
-    let site;
-
-    fetchModelById(req.params.site_id, Site)
-      .then((model) => {
-        if (!model) { throw 404; }
-        site = model;
-        return siteAuthorizer.findOne(req.user, site);
-      })
-      .then(() => Build.findAll({
-        attributes: ['id'],
-        where: { site: site.id },
-        order: [['createdAt', 'desc']],
-        limit: 100,
-      }))
-      .then(builds => buildSerializer.serialize(builds))
-      .then(buildJSON => res.json(buildJSON))
-      .catch(res.error);
+module.exports = wrapHandlers({
```

Comment:
Could you use `wrapHandlers` function where the controllers are associated to the route?  Would that make them less nested and composable?

Reply:
one could use `wrapHandlers` or `wrapHandler` in this way. For example:
```
router.post('/somepath', wrapHandler(Controller.action));
```
or
```
const Controller = wrapHandlers(require('./controller));
router.post('/somepath', Controller.action);
```

Express has built in functionality for handling errors, in that if you add an error handler via `app.use(someErrorHandler)` where `someErrorHandler` is a function with 4 arguments, when an error is thrown or  passed to `next(error)` in middleware, the error handlers will be called in order. However, this only works for synchronous controller actions. For async ones, we need to do a little more work and that is where `wrapHandler` comes in. All it does take an async function, add a `catch` to it and then call `next(error)`. Otherwise we would have to explicitly do this ourselves for each one with `catch` or `try/catch`.

Function definition:
```
function wrapHandler(fn) {
  // calls `next(error)`
  return (...args) => fn(...args).catch(args[2]);
}
```

This works because the third argument passed to the handler is `next`, we could make this more explicit:
```
function wrapHandler(fn) {
  return (req, res, next) => fn(req, res).catch(next);
}
```

Reply:
I updated the function definition as above.

Reply:
Sounds good.